app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala

```diff
@@ -766,6 +791,64 @@ class SectionRenderingService(
       )
   }
 
+  private def findFileId(ei: ExtraInfo, formComponent: FormComponent): FileId =
+    ei.envelope
+      .find(formComponent.modelComponentId)
+      .map(_.fileId)
+      .getOrElse(ei.envelope.mapping.fileIdFor(formComponent.id))
+
+  private def isVisible(formComponent: FormComponent, ei: ExtraInfo): Boolean = formComponent.includeIf.fold(true) {
+    includeIf =>
+      ei.formModelOptics.formModelVisibilityOptics.evalIncludeIfExpr(includeIf, None)
+  }
+
+  private def htmlForUpscan(
+    renderUnit: RenderUnit,
+    formTemplateId: FormTemplateId,
+    ei: ExtraInfo,
+    validationResult: ValidationResult,
+    upscanInitiate: UpscanInitiate
+  )(implicit
+    messages: Messages,
+    l: LangADT,
+    sse: SmartStringEvaluator
+  ): Html =
+    renderUnit.fold { case RenderUnit.Pure(formComponent) =>
+      if (!isVisible(formComponent, ei) || formComponent.onlyShowOnSummary) {
+        HtmlFormat.empty
+      } else {
+        formComponent.`type` match {
+          case FileUpload(fileUploadProvider) =>
+            val fileId: FileId = findFileId(ei, formComponent)
+            fileUploadProvider match {
+              case FileUploadProvider.Upscan =>
+                val upscanInitiateResponse: UpscanInitiateResponse = upscanInitiate.get(formComponent.id)
+                val hiddenFields: List[Html] = upscanInitiateResponse.uploadRequest.fields.toList.map {
+                  case (name, value) =>
+                    html.form.snippets.hidden(name, value)
+                }
+                val fileUploadName = "file"
+                val attributes = Map("upscan" -> "upscan")
+                htmlForFileUpload(
```

Comment:
I think we should render the final form right here ie. `formWithoutCSRF`. That form should be bare form with only `hiddenFields` and nothing else.

We should also capture metadata about this form. ie. `(fcId: FormComponentId, htmlId: String)`. We will need this once we render main form `gf-form`